### PR TITLE
[ONNX] Extend ONNX frontend with SoftmaxCrossEntropyLoss function #20547

### LINKED ISSUE #20547: [Good First Issue]: Extend ONNX Frontend with Function SoftmaxCrossEntropyLoss
### Context

Neural networks are graphs consisting of nodes called operators. Each operator corresponds to a mathematical function, usually described in framework's documentation or an AI standard, such as ONNX.
OpenVINO ONNX Frontend is a component responsible for working with ONNX graphs and requires implementation of different ONNX operators in order to use ONNX models.
This task requires extending OpenVINO ONNX Frontend with Function SoftmaxCrossEntropyLoss.													
Necessary help will be provided by ONNX Fronted team.

### What needs to be done?

Operator details can be found in [ONNX Operators](https://github.com/onnx/onnx/blob/main/docs/Operators.md#SoftmaxCrossEntropyLoss)
More details can be found in [ONNX Changelog](https://github.com/onnx/onnx/blob/main/docs/Changelog.md#SoftmaxCrossEntropyLoss-12)
1. Create `.hpp` and `.cpp` files for *Windows [here](https://github.com/openvinotoolkit/openvino/tree/master/src/frontends/onnx/frontend/src/op)													
2. Prepare an implementation of this operator in form of a function. SoftmaxCrossEntropyLoss-12 should be placed in [opset 1 namespace](https://github.com/openvinotoolkit/openvino/blob/master/src/frontends/onnx/frontend/src/op/unique.hpp#L13). SoftmaxCrossEntropyLoss-13 should be placed in [opset 13 namespace](https://github.com/openvinotoolkit/openvino/blob/master/src/frontends/onnx/frontend/src/op/softmax.cpp#L50).
3. Register the function in [ops_bridge.cpp](https://github.com/openvinotoolkit/openvino/blob/master/src/frontends/onnx/frontend/src/ops_bridge.cpp#L554-L560) while keeping alphabetical order
4. Create test model(s) in [ONNX models directory](https://github.com/openvinotoolkit/openvino/tree/master/src/frontends/onnx/tests/models/). OpenVINO test infrastructure then converts prototxt files to ONNX models - you will use those models later in tests
5. Add tests covering all use cases [here](https://github.com/openvinotoolkit/openvino/blob/master/src/frontends/onnx/tests/onnx_import.in.cpp)
6. Check [Python xfailed tests](https://github.com/openvinotoolkit/openvino/blob/master/src/frontends/onnx/tests/tests_python/test_backend.py) to find a test marked as a xfailed for added functionality. If any exist - remove corresponding lines and try to verify by using cmdline "python -m pytest -k name_of_test".
More details in [adding operators to ONNX Frontend guide](https://github.com/openvinotoolkit/openvino/blob/master/src/frontends/onnx/docs/how_to_add_op.md)

### Example Pull Requests

- https://github.com/openvinotoolkit/openvino/pull/15076
- https://github.com/openvinotoolkit/openvino/pull/14936
- https://github.com/openvinotoolkit/openvino/pull/12510
- https://github.com/openvinotoolkit/openvino/pull/19428

### Resources

- [What is OpenVINO?](https://github.com/openvinotoolkit/openvino#what-is-openvino-toolkit)
- [Contribution guide](https://github.com/openvinotoolkit/openvino/blob/master/CONTRIBUTING.md)
- [Blog post on contributing to OpenVINO](https://github.com/openvinotoolkit/openvino/blob/master/CONTRIBUTING.md)
- [User documentation](https://docs.openvino.ai/)
- Guides on [ONNX Frontend architecture](https://github.com/openvinotoolkit/openvino/blob/master/src/frontends/onnx/docs/architecture.md) and [ONNX Frontend tests](https://github.com/openvinotoolkit/openvino/blob/master/src/frontends/onnx/docs/tests.md) are also available
- You can also see a [blog post on contributing to OpenVINO](https://medium.com/openvino-toolkit/how-to-contribute-to-an-ai-open-source-project-c741f48e009e)

### Contact points

@gkrivor

### Ticket

_No response_

### PR DESCRIPTION
**Pull request by tanishka321**
### Details:
Extend ONNX frontend with SoftmaxCrossEntropyLoss function

### Tickets:
 Closes #20547

### CODE REVIEW

File: src/frontends/onnx/frontend/src/op/softmax_cross_entropy_loss.cpp [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+    const auto labels = node.get_ng_inputs().at(1);
+    
+    const auto axis = node.get_attribute_value<int64_t>("axis", 1);
+
+    std::shared_ptr<ngraph::Node> result;
+    result = onnx_softmax_cross_entropy_loss(logits, labels, axis);
+
+    return {result};
+}
+}  // namespace set_1
```
Comment (mitruska) [inline src/frontends/onnx/frontend/src/op/softmax_cross_entropy_loss.cpp:45]: This code is repeated 3 times for each namespace (set_1, set_11, set_13). It was a good direction to create common helper function `onnx_softmax_cross_entropy_loss` and reuse it, but more code can be unified. In fact there is no difference between the versions, just bfloat16 support has been added to the ONNX spec, so the import code can be shared between versions.
https://onnx.ai/onnx/operators/text_diff_SoftmaxCrossEntropyLoss_12_13.html

File: src/frontends/onnx/frontend/src/op/softmax_cross_entropy_loss.cpp [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+
+    return std::make_shared<default_opset::Reshape>(result, logits_shape, special_zero);
+}
+}  // namespace
+
+namespace op {
+namespace set_1 {
+OutputVector softmax_cross_entropy_loss(const Node& node) {
+    const auto logits = node.get_ng_inputs().at(0);
+    const auto labels = node.get_ng_inputs().at(1);
```
Comment (mitruska) [inline src/frontends/onnx/frontend/src/op/softmax_cross_entropy_loss.cpp:36]: It's better to call `node.get_ng_inputs()` once and assign it to a variable, then take required input by index.

File: src/frontends/onnx/tests/onnx_import.in.cpp [hunk 3]
```diff
@@ -6247,3 +6247,32 @@ OPENVINO_TEST(${BACKEND_NAME}, onnx_model_gelu_float_tanh) {
...
+
+OPENVINO_TEST(${BACKEND_NAME}, onnx_model_softmax_cross_entropy_loss_int) {
+    auto function = onnx_import::import_onnx_model(file_util::path_join(ov::test::utils::getExecutableDirectory(),
+                                                                        SERIALIZED_ZOO,
+                                                                        "onnx/softmax_cross_entropy_loss_int.onnx"));
+
+    auto test_case = ov::test::TestCase(function, s_device);
+    test_case.add_input<int64_t>(Shape{2}, {1, 2});
+    test_case.add_input<int64_t>(Shape{2}, {0, 2});
+    test_case.add_expected_output<bool>(Shape{2}, {false, true});
```
Comment (mitruska) [inline src/frontends/onnx/tests/onnx_import.in.cpp:6259]: Why the type of the expected output values is `bool`?
According to the ONNX [SoftmaxCrossEntropyLoss](https://onnx.ai/onnx/operators/onnx__SoftmaxCrossEntropyLoss.html#outputs) specification, it should be floating-point precision.

File: src/frontends/onnx/frontend/src/op/softmax_cross_entropy_loss.hpp [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+
+#include "ngraph/node.hpp"
+#include "onnx_import/core/node.hpp"
+
+namespace ngraph {
+namespace onnx_import {
+namespace op {
+namespace set_1 {
+
+OutputVector softmax_cross_entropy_loss(const Node& node);
```
Comment (mitruska) [inline src/frontends/onnx/frontend/src/op/softmax_cross_entropy_loss.hpp:18]: Please sync with the latest master. ONNX FE has been slightly refactored. The changes (namespaces, headers) need to be aligned with the rest of the frontends/onnx files.

Any of the onnx frontend op file can be taken as an example to follow, like:

[/src/frontends/onnx/frontend/src/op/leaky_relu.hpp](https://github.com/openvinotoolkit/openvino/blob/master/src/frontends/onnx/frontend/src/op/leaky_relu.hpp)
[/src/frontends/onnx/frontend/src/op/leaky_relu.cpp](https://github.com/openvinotoolkit/openvino/blob/master/src/frontends/onnx/frontend/src/op/leaky_relu.cpp)

**Review by mitruska — CHANGES_REQUESTED** [review]
Hello @tanishka321, thank you for your contribution.

Decomposition subgraph can contain only extisting OpenVINO operators,
it's not possible to create `default_opset::SoftmaxCrossEntropyLoss` as such class doesn't exists (so such code can't be compiled). SoftmaxCrossEntropyLoss operation is not available as OpenVINO primitive operator ([list of ov ops](https://docs.openvino.ai/2023.3/openvino_docs_ops_opset13.html)), that's why the ONNX Function expand mechanism is used.

@gkrivor As ONNX SoftmaxCrossEntropyLoss is training related operator, I would suggest to verify it's applicability for custom decomposition in OpenVINO ONNX FE.
